staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go

```diff
@@ -170,3 +174,19 @@ func compressUsername(username string) string {
 		return "other"
 	}
 }
+
+func authenticatorIDFromResponse(response *authenticator.Response) string {
+	if response == nil || response.AuthenticatorName == "" {
+		return unknownAuthenticator
+	}
+	return response.AuthenticatorName
+}
+
+func authenticatorIDFromError(err error) string {
```

Comment:
The error can be an aggregated error from multiple authenticators so I am not sure this really makes sense.

Reply:
Yes, the [union](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go) authenticator makes this a bit tricky.

However, 
1. When union authentication succeeds, we can correctly attribute the successful authentication action to the type that performed it.
2. In the case of a failure (and when FailOnError is true), we can correctly attribute failed authentication actions.
3. The part that is a bit fuzzy is when all authentication handlers within a union fail. In this case, we have a couple of options:
   a. Since the authenticators that make-up the union will increment their respective metrics, we 
    can can leave the union authenticator as is. Thus, the final metric (produced by union) will 
    be classified as "unknown".
   b. Classify the final metric as "union".